lua/bufferline/events.lua

```diff
@@ -0,0 +1,248 @@
+      schedule(function() render.update(true) end)
+    end,
+    group = augroup_render,
+  })
+
+  create_autocmd('TermOpen', {
+    callback = function() defer_fn(function() render.update(true) end, 500) end,
+    group = augroup_render,
+  })
+
+  create_autocmd('User', {
+    callback = function()
+      -- We're allowed to use relative paths for buffers iff there are no tabpages
+      -- or windows with a local directory (:tcd and :lcd)
+      local use_relative_file_paths = true
+
+      -- PERF: I didn't import `haslocaldir` since it is only used here (just before calling `:mksession` and exiting vim)
+      local haslocaldir = vim.fn.haslocaldir --- @type function
+
+      for tabnr, tabpage in ipairs(list_tabpages()) do
+        if not use_relative_file_paths or haslocaldir(-1, tabnr) == 1 then
+          use_relative_file_paths = false
+          break
+        end
+
+        for _, win in ipairs(tabpage_list_wins(tabpage)) do
+          if haslocaldir(win, tabnr) == 1 then
+            use_relative_file_paths = false
+            break
+          end
+        end
+      end
+
+      local buffers = {}
+      for _, bufnr in ipairs(state.buffers) do
+        local name = buf_get_name(bufnr)
+        if use_relative_file_paths then
+          name = utils.relative(name)
+        end
+
+        -- escape quotes
+        table_insert(buffers, {name = name, pinned =  state.get_buffer_data(bufnr).pinned})
+      end
+
+      vim.g.Bufferline__session_restore = "lua require'bufferline.state'.restore_buffers " ..
+        vim.inspect(buffers, {newline = ' ', indent = ''})
+    end,
+    group = augroup_misc,
+    pattern = 'SessionSavePre',
+  })
+
+  create_autocmd('WinClosed', {
+    callback = schedule_wrap(render.update),
+    group = augroup_render,
+  })
+
+  vim.schedule(function()
+    dictwatcherdel()
+    vim.cmd [[
+      " Must be global -_-
+      function! BufferlineOnOptionChanged(dict) abort
+        call luaeval("require'bufferline.events'.on_option_changed(_A)", a:dict)
+      endfunction
+
+      " Must be global -_-
+      function! BufferlineDictChanged(dict, key, changes) abort
+        silent! call dictwatcherdel(a:changes.old, '*', 'BufferlineOnOptionChanged')
+        call dictwatcheradd(a:changes.new, '*', 'BufferlineOnOptionChanged')
+        call BufferlineOnOptionChanged(a:changes.new)
+      endfunction
+
+      call dictwatcheradd(g:, 'bufferline', 'BufferlineDictChanged')
+      call dictwatcheradd(g:bufferline, '*', 'BufferlineOnOptionChanged')
+    ]]
+  end)
+
+  render.update()
+  enabled = true
+end
+
+--- What to do when the user configuration changes
+--- @param user_config table
+--- @return nil
+function events.on_option_changed(user_config)
+  options.setup(user_config) -- NOTE: must be first `setup` called here
+  highlight.setup()
+  JumpMode.set_letters(options.letters())
+
+  -- Don't jump-start barbar if it is disabled
+  if enabled then
+    render.update()
```

Comment:
This is a bug fix from the old `render.on_option_changed` implementation. If you ran `:BarbarDisable` and then did `g:bufferline.foo = 'bar'`, barbar would suddenly spring back to life. Now, you have to call `:BarbarEnable` again first.